packages/engine/paima-funnel/src/funnels/carp/funnel.ts

```diff
@@ -0,0 +1,269 @@
+import {
+  ChainDataExtensionType,
+  DEFAULT_FUNNEL_TIMEOUT,
+  ENV,
+  delay,
+  doLog,
+  logError,
+  timeout,
+} from '@paima/utils';
+import type {
+  ReadPresyncDataFrom,
+  ChainData,
+  ChainDataExtension,
+  ChainDataExtensionCardanoDelegation,
+  ChainFunnel,
+  PresyncChainData,
+  ChainDataExtensionDatum,
+} from '@paima/runtime';
+import { composeChainData, groupCdeData } from '../../utils';
+import { BaseFunnel } from '../BaseFunnel';
+import type { FunnelSharedData } from '../BaseFunnel';
+import type { PoolClient } from 'pg';
+import getCdePoolData from '../../cde/cardanoPool';
+import { Network } from '@paima/utils/src/constants';
+import { query, getErrorResponse } from '@dcspark/carp-client/client/src/index';
+import { Routes } from '@dcspark/carp-client/shared/routes';
+
+const confirmationDepth = '10';
+
+function knownTime(): number {
+  switch (ENV.CARDANO_NETWORK) {
+    case 'preview':
+      return 1666656000;
+    case 'preprod':
+      return 1666656000;
+    case 'mainnet':
+      return 1666656000;
+    default:
+      throw new Error('unknown cardano network');
+  }
+}
+
+function timestampToAbsoluteSlot(timestamp: number): number {
```

Comment:
do we care about byron though? With delegation it doesn't make sense afaik. Is there a case we we may need to consider pre-shelley blocks?

Reply:
By `AbsoluteSlot` I mean the cardano slot relative to the start of the chain, and not to the epoch. Which sometimes is also just called `slot` (like in carp).

`knownTime`  (now called `knownShelleyTime`) is the unix timestamp of the first block of shelley, which in the case of preview for example is also the genesis, but for mainnet is now.

Reply:
Also, answering my own question

>  Is there a case we we may need to consider pre-shelley blocks?

This is not used during the presync, it only matters during the sync, in order to map the EVM block to a unique slot in cardano, so that wouldn't be a problem anyway.